SDL3 Support

Fancy2209 opened the pull request:
Current work to get Lime working on SDL3

Fixes: #1848

Fancy2209 committed:
Updated submodule to SDL 3.2.8

Fancy2209 committed:
Commit current progress

Fancy2209 committed:
Fix typo

Fancy2209 committed:
Update sdl3-files.xml

ghost commented:
the thing is sdl module needs to switched with sdl3 and vulkan implementation needed in lime and SDL android java codes needs to be updated too

Fancy2209 commented:
> the thing is sdl module needs to switched with sdl3 and vulkan implementation needed in lime and SDL android java codes needs to be updated too

I did update the module to SDL3.
SDL still supports OpenGL, no need for Vulkan.
I will update Android Java when I get this working on desktoo

ghost commented:
> I did update the module to SDL3.

I meant to say SDL2 needs to be removed from the lime and fully switched with SDL3

Fancy2209 commented:
> > I did update the module to SDL3.
> 
> I meant to say SDL2 needs to be removed from the lime and fully switched with SDL3

It does not? I don't see why SDL2 shouldn't be an option.

Fancy2209 committed:
Add config for linux, still missing files on sdl3-files.xml

Fancy2209 committed:
It builds now!

Fancy2209 committed:
Fix SDL_Init

EliteMasterEric commented:
Interesting that the proposal is to have both available

Maybe the SDL2 can be available, and used by default, and deprecated later as Lime's SDL3 implementation is proven to be stable.

Fancy2209 commented:
> Interesting that the proposal is to have both available
> 
> Maybe the SDL2 can be available, and used by default, and deprecated later as Lime's SDL3 implementation is proven to be stable.

It was mostly because mojoAL doesn't have a SDL3 version yet, and from what I was reading on the build files it's used for Audio on the Switch

ACrazyTown commented:
It'a also used instead of OpenAL Soft when static linking (due to licensing conflicts) I believe

Fancy2209 committed:
Merge branch 'openfl:develop' into SDL3

Fancy2209 committed:
Fix SDL Include

Fancy2209 committed:
Fix Android build errors

Fancy2209 committed:
Fix Windows SDL3

Fancy2209 committed:
Actually fix SDLWindow.cpp

Fancy2209 committed:
Update SDLWindow.cpp

Fancy2209 committed:
Update SDLJoystick.cpp

Fancy2209 committed:
Actually actually fix windows

Fancy2209 committed:
Update Build.xml

Fancy2209 commented:
I accidentally pulled develop into tbe branch, not sure what to do about that.

Fancy2209 committed:
Update Build.xml

Fancy2209 committed:
Add AVFoundation to list of linked libs

Fancy2209 committed:
Link CoreHaptics on macOS

Fancy2209 committed:
Atualizar o config-windows-x86.h

Fancy2209 committed:
Atualizar o config-windows-x86_64.h

ACrazyTown commented:
What's the status of this?

Fancy2209 commented:
> What's the status of this?

Builds on Linux, doesn't build on everything else,  I was having issues with asset manifests due to Haxe's JSON parser locally

tobil4sk commented on project/src/backend/sdl3/SDLSystem.cpp:
```diff
@@ -0,0 +1,997 @@
+
+		#ifndef HX_WINDOWS
+
+		nmem = SDL_ReadIO (stream ? (SDL_IOStream*)stream->handle : NULL, ptr, size);
```
`size` here is the size of a unit to read, but here SDL_ReadIO is taking the length in bytes of the content to read. I think we should instead pass `size * count`. This is what caused asset manifest json files to be read in incorrectly.

Fancy2209 replied:
Have you been able to confirm this fixes the issue?

tobil4sk replied:
Yes, this has fixed the json loading for me. However, after solving this I now get a crash during png loading which seems like it could have a similar cause.

Note: the fwrite method below also uses `size` incorrectly like this, though this also doesn't fix the png problem.

Fancy2209 replied:
> Yes, this has fixed the json loading for me. However, after solving this I now get a crash during png loading which seems like it could have a similar cause.
> 
> Note: the fwrite method below also uses `size` incorrectly like this, though this also doesn't fix the png problem.

Last commit fixes it (I think), I had missed the output is now meant to be divided by size too

Fancy2209 committed:
Fix fread and fwrite partially

ACrazyTown commented:
I've gotten this to build on Windows!

![Screenshot 2025-03-28 141614](https://github.com/user-attachments/assets/69996d07-29f6-4e7e-971c-4f375b577cee)

Here's the commit from my fork: https://github.com/ACrazyTown/lime/commit/9993b5cb29216fb81cf0534d9cfad39a3d13b436. I would've done a pull request to your fork but GitHub won't let me for whatever reason

Fancy2209 commented:
> I've gotten this to build on Windows!
> 
> ![Screenshot 2025-03-28 141614](https://private-user-images.githubusercontent.com/47027981/428044188-69996d07-29f6-4e7e-971c-4f375b577cee.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMxNzIzMDUsIm5iZiI6MTc0MzE3MjAwNSwicGF0aCI6Ii80NzAyNzk4MS80MjgwNDQxODgtNjk5OTZkMDctMjlmNi00ZTdlLTk3MWMtNGYzNzViNTc3Y2VlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzI4VDE0MjY0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA2MDA5ZTc0YWE1ZjE1MTVjY2RmMTE0NGQ3ODBhNzc1ZWJiNWU0ZjBlOGU2NzU3YzYzN2E2NmM5MWYwMmJiNGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.03MP7ddnwGKZYS70zMzSsTsI5-x5uo0C9b4gIR0AE_0)
> 
> Here's the commit from my fork: [ACrazyTown@9993b5c](https://github.com/ACrazyTown/lime/commit/9993b5cb29216fb81cf0534d9cfad39a3d13b436). I would've done a pull request to your fork but GitHub won't let me for whatever reason

I can just commit it and it should retain you as the author, thank you!
Will try fixing IO outside windows

Fancy2209 committed:
fix fread and fwrite

Fancy2209 committed:
Fix formatting

tobil4sk commented on project/src/backend/sdl3/SDLSystem.cpp:
```diff
@@ -0,0 +1,1003 @@
+
+
+	FILE* FILE_HANDLE::getFile () {
+		return (FILE*)handle;
```
The SDL_IOStream is now an opaque struct, so this cast no longer safe and causes segmentation faults on Linux when the FILE* is passed to png functions. Accessing it via the property solves this problem:

```suggestion
		SDL_PropertiesID properties = SDL_GetIOProperties((SDL_IOStream*)handle);
		return (FILE*)SDL_GetPointerProperty(properties, SDL_PROP_IOSTREAM_STDIO_FILE_POINTER, NULL);
```
> The internal FILE * is available through a standard SDL_IOStream property, for streams made through SDL_IOFromFile() that use stdio behind the scenes; apps use this pointer at their own risk and should make sure that SDL and the app are using the same C runtime.

https://wiki.libsdl.org/SDL3/README/migration#sdl_rwopsh

`SDL_IOFromFile` sets this property, see also:
https://wiki.libsdl.org/SDL3/SDL_IOFromFile#remarks

We may need separate logic for android and possibly windows

tobil4sk replied:
> We may need separate logic for android and possibly windows

Compare to the sdl2 implementation:
https://github.com/Fancy2209/lime/blob/d74b0cbcf972b3287c02ef5a357543275e13a7b7/project/src/backend/sdl/SDLSystem.cpp#L602-L641

Fancy2209 committed:
Update project/src/backend/sdl3/SDLSystem.cpp

Co-authored-by: tobil4sk <tobil4sk@outlook.com>

Fancy2209 committed:
Fix building on Windows

Co-authored-by: ACrazyTown <luigibros626@gmail.com>

Fancy2209 committed:
Fix whitespace on SDLSystem.cpp

Fancy2209 committed:
Update Build.xml

Fancy2209 committed:
Fix Android, iOS and macOS hopefully

Fancy2209 committed:
Remove UIKit from macOS

Fancy2209 committed:
Add missing frameworks for iOS and TVOS

Fancy2209 committed:
Fix JNI.cpp

Fancy2209 committed:
Add missing file in SDL3

Fancy2209 committed:
Update Build.xml

Fancy2209 committed:
Update Build.xml

Fancy2209 committed:
Update openal-files.xml

Fancy2209 committed:
Fix paths to renamed files in SDL3

Fancy2209 committed:
SDL_uikit_main.c no longer exists

Fancy2209 committed:
SDL_android_main.c is no longer a thing

Fancy2209 committed:
Update libsdl Java and fix iOS

Fancy2209 committed:
Don't build HelloWorld on Windows

Fancy2209 committed:
Use SDL2 on WinRT, Switch, or if mojoAL is used

Fancy2209 committed:
Install SDL3 with brew on iOS Ci

Fancy2209 committed:
Include SDL3 Headers in template

Fancy2209 committed:
Stop installing SDL3 with brew on iOS Ci

Fancy2209 committed:
Print error in SVG

Fancy2209 committed:
Improve SVG Error Log

Fancy2209 committed:
Fix headers included in ios template

Fancy2209 committed:
Update svg.n

Fancy2209 committed:
Rebuild svg.n on Ci

Fancy2209 committed:
Install deps for svg.n

Fancy2209 committed:
Fix FILE_HANDLE::getFile() on Windows

woops

Fancy2209 committed:
Add Android file support back

Fancy2209 committed:
Add missing semicolon

Fancy2209 committed:
Add missing iOS Deps

Fancy2209 committed:
Fix ios pbxproj

Fancy2209 commented:
That was a lot of work, wew
Finally all targets pass Ci and now this is ready for proper reviewing and better testing

Fancy2209 committed:
fix GetDisplay

Fancy2209 committed:
Fix Android

EliteMasterEric reviewed (approved):
We have been reviewing this pull request internally and have found no major issues so far, only a couple missing graphics (which should be trivially resolved on our end). Performance seems similar and it would be great to have the new features introduced once this is merged.

Fancy2209 commented:
> We have been reviewing this pull request internally and have found no major issues so far, only a couple missing graphics (which should be trivially resolved on our end). Performance seems similar and it would be great to have the new features introduced once this is merged.

Good to know!

About missing graphics, I've had a few FileIO issues, but I thought I had (finally) fully fixed them, could you elaborate one what you mean? I just wanna make sure I didn't break something

Kade-github commented:
Theres only one function in this PR that I needed to get, which was SDL_GetTicksNS(), which is similar to the function being exposed as "System::GetTimer();" I implemented this as "System::GetTimerNS();" but I am wondering of other SDL3 only features have yet to be exposed? Otherwise this PR looks good :D

EliteMasterEric commented:
> > We have been reviewing this pull request internally and have found no major issues so far, only a couple missing graphics (which should be trivially resolved on our end). Performance seems similar and it would be great to have the new features introduced once this is merged.
> 
> Good to know!
> 
> About missing graphics, I've had a few FileIO issues, but I thought I had (finally) fully fixed them, could you elaborate one what you mean? I just wanna make sure I didn't break something

After more testing I found that the missing graphic was related to how I was handling SDL timestamps, and resolving the issue fixed the graphic. There are no FileIO issues that I am aware of.

Fancy2209 commented:
> Theres only one function in this PR that I needed to get, which was SDL_GetTicksNS(), which is similar to the function being exposed as "System::GetTimer();" I implemented this as "System::GetTimerNS();" but I am wondering of other SDL3 only features have yet to be exposed? Otherwise this PR looks good :D

Oh yeah definitely, I at the very least want to 
- expose SDL_GetTicksNS like y'all did on Funkin
- Allow Windows to optionally support for transparent backgrounds (I already figured this one out just waiting on this to get merged before PRing it)
- replace file dialogs with the SDL Dialog API when using SDL3
- Expose SDL_GetSystemTheme

I also know SDL3 adds a camera API who would allow us to implement Camera support, same thing with the System Tray API.
I also need to check what we need to change for Clipboard to handle the fact it can get more than just text now.

The only thing currently exposed is that enabling high DPI should also work on windows now, the above is only my wishlist/to work on list after this one gets merged.

If you wanna see everything SDL3 adds this is a pretty good list, though I noticed it doesn't mention some things like SDL_WINDOW_TRANSPARENT
https://wiki.libsdl.org/SDL3/NewFeatures

joshtynjala commented:
> I am wondering of other SDL3 only features have yet to be exposed

We've asked Fancy to wait on exposing any of SDL3's new features. To be clear, we definitely plan to take advantage of many of those new features in Lime and OpenFL because there are a lot of very nice things in SDL3. However, in terms of testing and review, it's much easier if we can focus on the upgrade with the same set of features that already existed in SDL2. After we merge this foundational SDL3 support, then we can start having fun with the new stuff.

Kade-github commented:
> > I am wondering of other SDL3 only features have yet to be exposed
> >
> To be clear, we definitely plan to take advantage of many of those new features in Lime and OpenFL because there are a lot of very nice things in SDL3. 

Awesome! Can't wait :D

EliteMasterEric commented:
It will also be great to start seeing changes in OpenFL as well, such as replacing this line with the proper nanosecond timestamp call:

https://github.com/openfl/openfl/blob/782627b89de76ff6ca85c267f8e36e2e09e1dc92/src/openfl/display/FPS.hx#L54

(i.e. finally resolving #1569 )

dimensionscape commented:
> It will also be great to start seeing changes in OpenFL as well, such as replacing this line with the proper nanosecond timestamp call:
> 
> https://github.com/openfl/openfl/blob/782627b89de76ff6ca85c267f8e36e2e09e1dc92/src/openfl/display/FPS.hx#L54
> 
> (i.e. finally resolving #1569 )

That line only affects the flash target... and does not actually solve anything. We also can't use anything from our native libraries in flash at all.

WIth that said, the main loop has been overhauled for exact preciseness with better support for vsync (and fallback for hardware that doesnt support vsync). Most of that work is about 95% complete and should make it's way into 8.3 over the coming months.

KarimAkra commented:
I tried this on Android and image files (PNG atleast) decoding seems to have broke? I did some logging and ended up with the assumption that the `lime::fread` method on `SDLSystem` isn't working properly (although it's literally the same thing in SDL3 migration docs... could be an internal issue with SDL..? Tho i found nothing on SDL issues)
I might be wrong but if that is the source of the issue then any C++ target other than Windows should have Image decoding issues and anything that uses `lime::fread` would be broken on these platforms

KarimAkra commented:
Also forgot to mention that trying to get displayMode is throwing a segfault for me on Android

Fancy2209 commented:
> Tho i found nothing on SDL issues) I might be wrong but if that is the source of the issue then any C++ target other than Windows should have Image decoding issues and anything that uses `lime::fread` would be broken on these platforms

I know it isn't because this whole PR was made on Linux... Very weird Android is broken

gepatto commented:
Just chiming in.
I've tested on a Raspberry Pi.
And it works with some adjustments to the build.xml
I have been working with a default sdl3-config file that one gets when compiling sdl3 with default settings.
It's a bit different from custom linux sdl3 config in this lime branch so I'm using a separate file for the Raspberry Pi.

dimensionscape commented:
> Just chiming in. I've tested on a Raspberry Pi. And it works with some adjustments to the build.xml I have been working with a default sdl3-config file that one gets when compiling sdl3 with default settings. It's a bit different from custom linux sdl3 config in this lime branch so I'm using a separate file for the Raspberry Pi.

Yay this is really important. I'm glad you chimed in I was just wondering how well this worked on Pi.

Fancy2209 commented:
I will close this PR and re open a version with a cleaner tree soon, will also take the time to try to wire mojoAL as it now supports SDL3

joshtynjala committed:
Merge branch '9.0.0-dev' into SDL3-mac

joshtynjala committed:
SDLGamepad: SDL3 rumble

joshtynjala commented:
Today, I took some time to check out this branch, merge in the latest from 9.0.0-dev, and do a little local testing.

The Feathers UI components-explorer sample builds and runs with neko, hl, and cpp targets on macOS. However, rendering on my HiDPI monitor is not quite right when I use`<window allow-high-dpi="true"/>`. Everything renders at the correct scale factor, but it's like the stage bounds reach beyond the window on all four sides. I suspect that this is related to the way that coordinates have changed in SDL3 to better support HiDPI.

For iOS, it builds and launches. However, I'm seeing only a white screen when I launch components-explorer. I haven't investigated deeper than that.

For Android, it builds, launches, but renders everything at a _super_ tiny scale factor. Probably a similar scaling issue with SDL3's changes to coordinates for HiDPI.

I'm planning to test on Windows and Linux later, but these are my initial observations on my main machine.

Fancy2209 commented:
> Today, I took some time to check out this branch, merge in the latest from 9.0.0-dev, and do a little local testing.
> 
> The Feathers UI components-explorer sample builds and runs with neko, hl, and cpp targets on macOS. However, rendering on my HiDPI monitor is not quite right when I use`<window allow-high-dpi="true"/>`. Everything renders at the correct scale factor, but it's like the stage bounds reach beyond the window on all four sides. I suspect that this is related to the way that coordinates have changed in SDL3 to better support HiDPI.
> 
> For iOS, it builds and launches. However, I'm seeing only a white screen when I launch components-explorer. I haven't investigated deeper than that.
> 
> For Android, it builds, launches, but renders everything at a _super_ tiny scale factor. Probably a similar scaling issue with SDL3's changes to coordinates for HiDPI.
> 
> I'm planning to test on Windows and Linux later, but these are my initial observations on my main machine.

That sounds like me using SDL_GetWindowSize() instead of SDL_GL_GetDrawableSize() somewhere

joshtynjala commented:
And I tested Windows and Linux this afternoon. I got sidetracked for a bit because I needed to upgrade from Visual Studio Community 2019 to 2022 to get SDL3 to compile on Windows. I was getting the following error:

> Cannot open include file: 'gameinput.h': No such file or directory

Upgrading to VS 2022 fixed it, for some reason. That's fine. I'm glad I'm not on an ancient version of VS anymore.

On Linux, rebuilding with hxcpp 4.3.2 didn't work, but I upgraded to latest hxcpp from GitHub and rebuild completed successfully. This is fine. By the time we release Lime 9, hxcpp should be updated on Haxelib anyway.

Anyway, on both Windows and Linux, I could compile and run the Feathers UI components-explorer sample. However, both rendered the window at half of the normal width and height. When I tried to set `allow-high-dpi="false"`, Windows and Linux still had a window at half width/height.

joshtynjala committed:
SDLApplication, SDLWindow: account for display scale and pixel density for high-dpi displays

joshtynjala committed:
SDLSystem: fix GetDisplay

the display id 0 is now considered an invalid display, and it seems like display ids aren't necessarily incrementing from 1, so check if name is null to determine if id is invalid instead of checking the value of GetNumDisplays()

joshtynjala committed:
SDLSystem: fix GetDisplay range error due to new starting display id

joshtynjala committed:
SDLWindow: fix resizing programmatically accounting for display content scale

MAJigsaw77 committed:
Android: remove png_set_sig_bytes to prevent PNG crash

Fancy2209 committed:
Merge pull request #1 from MAJigsaw77/patch-1

Android: remove png_set_sig_bytes to prevent PNG crash

ACrazyTown commented on project/src/backend/sdl3/SDLSystem.cpp:
```diff
@@ -0,0 +1,1041 @@
+			hl_dyn_setp (display, id_bounds, &hlt_dynobj, _bounds);
+
+			Rectangle safeAreaInsets;
+			Display::GetSafeAreaInsets(id, &safeAreaInsets);
```
Was getting a crash on macOS when targeting Hashlink, this fixes it:

```suggestion
			Display::GetSafeAreaInsets(id - 1, &safeAreaInsets);
```

joshtynjala committed:
SDLSystem: Fix display id on macOS

Co-authored-by: ACrazyTown <47027981+ACrazyTown@users.noreply.github.com>
